#  Assignment 01 - PyTorch Basics & Tensors

### Due: Monday, Jan 26, before class

Name(s): Chang Min<br>
Course: CSCI 357 - AI and Neural Networks<br>
Section: 01 - 1:00pm<br>
Semester: Spring 2026<br>
Instructor: Prof. King<br>

### References
* Some of the lab sections are based on the official tutorials at https://docs.pytorch.org/tutorials/beginner/colab

In [2]:
%%html
<style>
.todo-box {
    background-color: #fcf8e3;
    color: #000000;
    border-left: 6px solid #f0ad4e;
    padding: 10px;
    margin-bottom: 0;
}

.todo-box h3 {
    margin-top: 0;
}

.note-box {
    background-color: #d1ecf1;
    color: #000000;
    border-left: 6px solid #0c5460;
    padding: 10px;
    margin-bottom: 0;
}

.warning-box {
    background-color: #f8d7da;
    color: #000000;
    border-left: 6px solid #721c24;
    padding: 10px;
    margin-bottom: 0;
}

.section-box {
    background-color: rgb(125, 233, 150);
    color: #000000;
    border-left: 6px solid #155724;
    padding: 10px;
    margin-bottom: 0;
}
</style>

---

# Welcome

Welcome to your first assignment for CSCI 357! All assignments will follow a a fixed structure:
* **Background**: Essential context and theory for the week.
* **Lab**: A guided, hands-on exercise to get the "gears turning."
* **Exercises**: Harder, more challenging problems that require you to stretch your skills and apply the week's concepts to new scenarios. (NOTE: You will have to work through exercises that I may not have solutions for – but I am 100% certain they can be solved.)
* **Career Survey**: Analysis of real-world job postings and the evolving AI industry.
* **Society**: A reflection on mainstream media or research articles concerning the ethical and societal impact of AI.
* **AI Usage Disclosure**: A mandatory section where you carefully characterize your use of AI, including what worked, what failed, and how the AI assisted (or hindered) your learning.

Weekly assignments are designed to help you manage your time and spread out the work throughout the week, and not meant to be done at the last minute! **You will not get what you need out of this course if you wait until Sunday to work on the assignments in a single session and hope that AI will complete your work for you!** Don't cheat yourself out of learning!

**Assignments are released on Mondays after class. The entire assignment is due before the next class.**

Since this first week is a shortened week, and because you are just getting started, this will be a bit smaller and simpler than future assignments.


# Course Materials

Did you complete the following preparation steps for the course? Place an X next to each item you completed.<br>
- [x] Join Discord - https://discord.gg/NQkVDj7WD7
- [x] Read the syllabus, carefully.
- [x] Worked through the PDF `pre-HW01.pdf` on Moodle before starting here

In [3]:
%%html
<div class="todo-box">
<h3>TODO</h3>
<p>Complete the syllabus contract below by selecting the appropriate choices.</p>
</div>

# SYLLABUS CONTRACT

Select TRUE or FALSE for each question below by deleting choice that  so that it reads correct for you:

1. I read the syllabus<br>
TRUE 
2. I understand the course objectives and the learning outcomes<br>
TRUE 
3. I understand the course structure and the assignment breakdown<br>
TRUE 
4. I understand the grading policy and the expectations for collaboration<br>
TRUE 
5. I understand the course logistics and the expectations for attendance and participation<br>
TRUE 
6. I understand the course policies and the expectations for academic integrity<br>
TRUE 
7. I understand the AI requirements and expectations for this course<br>
TRUE 
8. I promise to commit to being mindful with my use of AI to only use to improve my learning and understanding and not cheat myself out of getting the most out of this course!<br>
TRUE 


In [4]:
# Some standard Python packages we'll use
import sys
import os

# Standard packages for the data scientist
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns

### A note about Colab

[Google Colab](https://colab.research.google.com/) is a free, cloud-based Jupyter environment that lets you run Python notebooks in your browser without installing anything locally. The **Colab Pro version is free for students**, which gives you access to **high-end GPUs** running in the cloud for faster training and experimentation.

### Can I sync a Google Drive folder with a folder on my laptop?

I personally switch between running locally on my machine or running in Colab. I use a folder locally on my Mac that is synced to Google Drive automatically, making the transition between both pretty seamless. I recommend [Google Drive for Desktop](https://support.google.com/a/users/answer/13022292?hl=en). It's free, it's Google's flagship product for this purpose, It does a nice job syncing folders between your local machine and your Drive account. Then, you get the best of both platforms, allowing you to go back and forth as needed.

> WARNING: If you are switching back and forth between Colab and your local machine, make sure you wait for synchronization to complete before opening files! Some notebooks will get large down the road.

Colab it's a great platform. It even integrates Gemini into the interface now, so you have access to a pretty good AI copilot. And, you won't get a simpler path to completing work in this course, as it alleviates you from needing to install anything. Or you could get the best of both workds simultaneously, and use the Google Colab plugin for VS Code to get the benefits of VSCode while running your code remotely in Colab.**


In [5]:
# Are we running from colab?
IN_COLAB = 'google.colab' in sys.modules

if IN_COLAB:
  print("Running in Google Colab")
else:
  print("Not running in Google Colab")

Running in Google Colab


In [6]:
%%html
<div class="section-box">
<h1></h1>
</div>

# <div class="section-box">Lab - Pytorch Basics & Tensors</div>

## Objectives

* Learn the fundamentals of PyTorch, with emphasis on tensors and their operations.
* Understand hardware acceleration

In [7]:
%%html
<div class="warning-box">
<h3>Many of these questions in the lab section will be very easy. AI should be able to solve most of them for you. Be mindful with how you use it. Treat these notebooks as your study guides. Don't become satisfied with mediocre crap in your answers. These notebooks will serve as references for prepping for tech interviews!</h3>
</div>

---
## PyTorch - a library for working with artificial neural networks in Python

[PyTorch](https://pytorch.org/) is a popular open-source machine learning library developed by [Meta's AI Research](https://ai.meta.com/research/) lab. It provides a flexible and intuitive framework for building and training neural networks. PyTorch is particularly known for its dynamic computational graph, which allows for more flexible model architectures and easier debugging compared to static graph frameworks.

Key features of PyTorch include:
- Dynamic computation graphs that can be modified on-the-fly
- Strong GPU acceleration support
- Extensive library of pre-built neural network layers and loss functions
- Automatic differentiation for gradient computation
- Seamless integration with Python's scientific computing ecosystem
- Active community and extensive documentation

PyTorch has become one of the most widely used deep learning frameworks, powering research and production applications across various domains including computer vision, natural language processing, and reinforcement learning.


## Section 1: Introduction to Tensors

At its core, the tensor in Pytorch is just like NumPy's `ndarray` (n-dimensional array), but with a couple of critical differences for neural nets:
1. Tensors can run on GPUs.
2. Tensors are optimized for automatic differentiation for building and training neural networks.

Let's dive a bit deeper into tensors.

A **tensor** is a specialized data structure for representing n-dimensional data. Yes, that means vectors for 1-d, matrices for 2-d, and so on. In PyTorch, we use **tensors** to encode all of the parameters of the model (i.e. the weights and biases), as well as the input and output of a model. All data you'll be working with must be encoded in tensors first! *Tensors are the most fundamental data structure in PyTorch, and you will use them in every asssignment. Therefore, it's important to really understand them.*



The standard package for PyTorch is `torch`:

In [8]:
import torch

In [9]:
print(f"PyTorch version: {torch.__version__}")

PyTorch version: 2.9.0+cu126


### 1.1 Creating Tensors

Tensors can be created in various ways. Let's explore a few:


In [10]:
# Directly from data
data = [[1, 2], [3, 4]]
x_data = torch.tensor(data)
print(f"Tensor from data:\n{x_data}\n")
print(f"x_data dtype is {x_data.dtype}")

# Or, directly from a NumPy array
np_array = np.array(data)
x_np = torch.from_numpy(np_array)
print(f"Tensor from NumPy:\n{x_np}\n")
np_array[0][0] = 0
print(f"Tensor from NumPy AFTER:\n{x_np}\n")

# Using constant or random values, while specifying the shape of the tensor
shape = (2, 3)
rand_tensor = torch.rand(shape)
ones_tensor = torch.ones(shape)
zeros_tensor = torch.zeros(shape)

# Printing them looks a lot like numpy printed data
print(f"Random Tensor: \n {rand_tensor} \n")
print(f"Ones Tensor: \n {ones_tensor} \n")
print(f"Zeros Tensor: \n {zeros_tensor}")

Tensor from data:
tensor([[1, 2],
        [3, 4]])

x_data dtype is torch.int64
Tensor from NumPy:
tensor([[1, 2],
        [3, 4]])

Tensor from NumPy AFTER:
tensor([[0, 2],
        [3, 4]])

Random Tensor: 
 tensor([[0.7224, 0.2255, 0.6822],
        [0.4311, 0.5701, 0.5571]]) 

Ones Tensor: 
 tensor([[1., 1., 1.],
        [1., 1., 1.]]) 

Zeros Tensor: 
 tensor([[0., 0., 0.],
        [0., 0., 0.]])


In [11]:

# We could generate a new matrix that mirrors the size of an existing matrix, and change the data type:
ones_tensor = torch.ones_like(x_data, dtype=torch.float)
print(f"original tensor: \n {x_data} \n")
print(f"ones tensor wth same shape: \n {ones_tensor} \n")

original tensor: 
 tensor([[1, 2],
        [3, 4]]) 

ones tensor wth same shape: 
 tensor([[1., 1.],
        [1., 1.]]) 



### 1.2 Tensor Attributes

As you can imagine, a tensor has a lot of attributes. Tensor attributes describe their shape, data type, and the device on which they are stored. (We'll discuss devices in more detail later in this lab.)


In [12]:
# We'll set up a tensor for demonstration purposes:
tensor = torch.rand(3, 4)

# Basic attributes
print(f"Shape: {tensor.shape}")
print(f"Datatype: {tensor.dtype}")
print(f"Device: {tensor.device}")
print(f"Number of dimensions: {tensor.ndim}")
print(f"\nTotal number of elements (numel): {tensor.numel()}")

# Additional attributes that will become useful down the road...
print(f"Requires gradient tracking: {tensor.requires_grad}")
print(f"Is a leaf tensor: {tensor.is_leaf}")
print(f"Is contiguous in memory: {tensor.is_contiguous()}")
print(f"Memory layout: {tensor.layout}")
print(f"Strides: {tensor.stride()}")

Shape: torch.Size([3, 4])
Datatype: torch.float32
Device: cpu
Number of dimensions: 2

Total number of elements (numel): 12
Requires gradient tracking: False
Is a leaf tensor: True
Is contiguous in memory: True
Memory layout: torch.strided
Strides: (4, 1)


In [13]:
%%html
<div class="todo-box">
<h3>TODO</h3>
</div>

### Question Set 1 - Tensors and Attributes

1. **Memory Sharing:** In Section 1.1, we see `torch.from_numpy()`. If you modify the original NumPy array after creating a tensor this way, does the tensor change? Why or why not?

**ANSWER:**<br>
Yes, the tensor changes after modifying the orignal NumPy as they share the same memory location for efficiency and speed. [source](https://perso.esiee.fr/~chierchg/deep-learning/tutorials/pytorch/pytorch-1.html)




2. **Device Awareness:** Looking at the attributes in Section 1.2, why is the `device` attribute critical for deep learning performance?

**ANSWER:**<br>The `device` attribute is important for deep learning performance as it determines whether computation acceleration is possible (e.g. using a GPU for matrix calculations).






3a. **Data Types:** What is the default data type for a tensor? If you create a tensor from a list of integers (like `[1, 2]`), what is the `dtype`? How does this differ from `torch.rand()`?

**ANSWER:**<br> The default data type for a tensor is `torch.float32` [source](https://docs.pytorch.org/tutorials/beginner/introyt/tensors_deeper_tutorial.html). If we create a tensor from a list of integers, the `dtype` will be `torch.int64`. This is different from `torch.rand()`, which uses the default data type of `torch.float32`.




3b. Write code to demonstrate the question above:

In [14]:
# ANSWER - 3b
int_list = [1, 2]
int_tensor = torch.tensor(int_list)
print(f"int_tensor has the dtype: {int_tensor.dtype}\n")

rand_tensor = torch.rand(1, 2)
print(f"rand_tensor has the dtype: {rand_tensor.dtype}\n")





int_tensor has the dtype: torch.int64

rand_tensor has the dtype: torch.float32




4. **Shape Manipulation:** Given the `tensor` variable created (3x4), use a single command to change its shape to 2x6, and show the contents.

In [15]:
# ANSWER - 4
manipulate_tensor = torch.rand(3,4)
print(f"Before: {manipulate_tensor}\n")
manipulate_tensor = manipulate_tensor.reshape(2,6)
print(f"After: {manipulate_tensor}\n")



Before: tensor([[0.6586, 0.1946, 0.6365, 0.7665],
        [0.1712, 0.7507, 0.7520, 0.0912],
        [0.6703, 0.5625, 0.4135, 0.1472]])

After: tensor([[0.6586, 0.1946, 0.6365, 0.7665, 0.1712, 0.7507],
        [0.7520, 0.0912, 0.6703, 0.5625, 0.4135, 0.1472]])



5. **Attribute Inspection:** Write code to check if the `ones_tensor` you created is stored on a GPU or CPU.

In [16]:
# ANSWER - 5
print(f"ones_tensor device: {ones_tensor.device}\n")



ones_tensor device: cpu



6. **Type Conversion:** Create a 2x3 tensor of ones with `dtype=torch.int32`, then convert it to `torch.float64` in a second step. Confirm the dtype and shape in your output.

In [17]:
# ANSWER - 6
convert_tensor = torch.ones((2,3), dtype=torch.int32)
print(f"original: {convert_tensor.dtype}, {convert_tensor.shape}\n")

convert_tensor = convert_tensor.to(torch.float64)
print(f"converted: {convert_tensor.dtype}, {convert_tensor.shape}\n")





original: torch.int32, torch.Size([2, 3])

converted: torch.float64, torch.Size([2, 3])



In [18]:
%%html
<div class="todo-box">
<h3>TODO</h3>
</div>

### Question Set 2 - Advanced Memory Layout & Strides

1. What is a stride in PyTorch? Describe the relationship between the shape and strides of a tensor.


**ANSWER:**<br>A stride in PyTorch describes how many steps/elements we need to pass in contiguous memory (which is how tensors are defined) to reach the next element of that specific dimension. The stride of a tensor is closely related to the dimensions/shape of a tensor as the stride determines how the tensor is read from memory. If the tensor is reshaped, the memory stays the same, but is read differently by changing the stride.









2. What is the difference between contiguous and non-contiguous memory layouts?

**ANSWER:**<br>Contiguous memory is when we have sequential elements also stored sequentially in memory. Non-contiguous memory is when we have sequential elements not being stored sequentially, so we need some way of jumping to these memory addresses that correspond to the next elements in the sequence.






3. Suppose you have a 2D tensor of shape `(M, N)`. When you call `.T` (transpose) on it, PyTorch does not move the data in memory. Instead, it creates a new "view" by modifying the **strides**.
   *   What are the strides of a contiguous tensor of shape `(3, 5)`?
   *   How do these strides change after applying `.T`?
   *   How does the `is_contiguous()` attribute change after transposing?


**ANSWER:**<br> The strides of a contiguous tensor of shape (3, 5) would be (5, 1). When we transpose it, the strides will flip to become (1, 5). `is_contiguous` will go from being `True` to `False`.







4. Let's explore the cost with respect to memory layout of transposing a large tensor. Write code to acheive the following:<br>
a. Create a large 2D tensor (e.g., 5000x7500).<br>
b. Transpose it to create a non-contiguous view.<br>
c. Use the `.stride()` and `.is_contiguous()` methods to inspect both the original and the transpose.<br>
d. Call `.contiguous()` on the transposed tensor. Inspect the strides again.<br><br>
Does the memory layout change? Hint: Use the `.data_ptr()` method to see if the memory address changes.

In [19]:
# ANSWER:

# Create a large tensor
large_tensor = torch.ones((5000,7500))


# Transpose it
large_tensor_t = large_tensor.T


# Check out stride and contiguity with the original and transpose
print(f"Original: {large_tensor.stride()}, {large_tensor.is_contiguous()}, {large_tensor.data_ptr()}")

print(f"Original: {large_tensor_t.stride()}, {large_tensor_t.is_contiguous()}, {large_tensor_t.data_ptr()}")


# Force contiguity
large_tensor_t = large_tensor_t.contiguous()

# Compare the pointers in your result. Forced contiguity should result in a different memory address (a copy).
print(f"New data pointers: {large_tensor.data_ptr()}, {large_tensor_t.data_ptr()}")

Original: (7500, 1), True, 139708960383040
Original: (1, 7500), False, 139708960383040
New data pointers: 139708960383040, 139708810379328


5. Do some research. You can see that there are a variety of operations on tensors that just change the "view" of the tensor, it doesn't actually copy or move any data in memory. PyTorch simply changes the **stride** and **shape** metadata to reflect the new view. This can save a massive amount of memory and computation time! Transposing a 1GB tensor costs almost zero extra memory!<br><br> Look into **Continguity**. If you did the above correct, you notice that transposing your tensor made it so that your memory was no longer stored in one contiguous block. Does this matter? HINT: Your answer should consider the memory layout of the tensor and the memory layout of the GPU.


**ANSWER:**<br>Yes, it matters if a tensor is contiguous or not because GPUs are optimized to read memory sequentially. If a tensor is not contiguous, we need to spend more time jumping around between memory addresses for operations. GPUs are optimized for coalesced memory access (servicing multiple logical memory addresses in a single physical memory address). [source1](https://hackernoon.com/pytorch-contiguous-tensor-optimization) [source2](https://modal.com/gpu-glossary/perf/memory-coalescing) [source3](https://www.arccompute.io/arc-blog/gpu-101-memory-hierarchy)







---
### 1.3 Arithmetic Operations with Tensors

Arithmetic operations with tensors in PyTorch have to be highly optimized. They are fundamental in deep learning and neural networks for several key reasons:

1. **Neural Network Computations**
   - Weight adjustments during training
   - Combining inputs with weights
   - Adding bias terms
   - Computing activation functions

2. **Feature Engineering**
   - Normalizing input data
   - Scaling features
   - Creating derived features

3. **Batch Processing**
   - Operating on multiple samples simultaneously
   - Efficient parallel computations on GPUs

4. **Mathematical Operations**
   - Matrix multiplication for layer transformations
   - Element-wise operations for activation functions
   - Computing loss functions
   - Gradient calculations during backpropagation

5. **Data Transformations**
   - Reshaping data for model input
   - Converting between different data representations
   - Applying mathematical transformations

There are over 1200 tensor operations, including transposing, indexing, slicing, mathematical operations,random sampling, linear algebra, and more.


#### Indexing and Slicing

PyTorch tensors follow standard NumPy-like indexing and slicing.


In [20]:
# Initialize the tensor with values 1, 2, 3, etc. not ones.
tensor = torch.arange(1,21).reshape([5,4])
print(tensor)
print('First row: ', tensor[0])
print('First column: ', tensor[:, 0])
print('First column: ', tensor[..., 0])
print('Last column:', tensor[..., -1])
tensor[:, 1] = 0
print(tensor)

tensor([[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12],
        [13, 14, 15, 16],
        [17, 18, 19, 20]])
First row:  tensor([1, 2, 3, 4])
First column:  tensor([ 1,  5,  9, 13, 17])
First column:  tensor([ 1,  5,  9, 13, 17])
Last column: tensor([ 4,  8, 12, 16, 20])
tensor([[ 1,  0,  3,  4],
        [ 5,  0,  7,  8],
        [ 9,  0, 11, 12],
        [13,  0, 15, 16],
        [17,  0, 19, 20]])


#### Joining tensors

You can use `torch.cat` to concatenate a sequence of
tensors along a given dimension. See also
[torch.stack](https://pytorch.org/docs/stable/generated/torch.stack.html),
another tensor joining operator that is subtly different from
`torch.cat`.


In [21]:
tensor = torch.arange(1,13).reshape(2,6)
print("Original tensor:\n", tensor)
t1 = torch.cat([tensor, tensor, tensor], dim=-1)
print(t1)

t4 = torch.cat([tensor, tensor, tensor], dim=0)
print(t4)
# Demonstrating torch.stack as well
# torch.cat concatenates tensors along an existing dimension.
# torch.stack joins a sequence of tensors along a new dimension.

# Let's stack three copies of the tensor along dim=0
t2 = torch.stack([tensor, tensor, tensor], dim=0)
print("\nStacked tensor shape:", t2.shape)
print("Stacked tensor (dim=0):\n", t2)

# You can also stack along another dimension, e.g., last dimension
t3 = torch.stack([tensor, tensor, tensor], dim=-1)
print("\nStacked tensor shape (dim=-1):", t3.shape)
print("Stacked tensor (dim=-1):\n", t3)



Original tensor:
 tensor([[ 1,  2,  3,  4,  5,  6],
        [ 7,  8,  9, 10, 11, 12]])
tensor([[ 1,  2,  3,  4,  5,  6,  1,  2,  3,  4,  5,  6,  1,  2,  3,  4,  5,  6],
        [ 7,  8,  9, 10, 11, 12,  7,  8,  9, 10, 11, 12,  7,  8,  9, 10, 11, 12]])
tensor([[ 1,  2,  3,  4,  5,  6],
        [ 7,  8,  9, 10, 11, 12],
        [ 1,  2,  3,  4,  5,  6],
        [ 7,  8,  9, 10, 11, 12],
        [ 1,  2,  3,  4,  5,  6],
        [ 7,  8,  9, 10, 11, 12]])

Stacked tensor shape: torch.Size([3, 2, 6])
Stacked tensor (dim=0):
 tensor([[[ 1,  2,  3,  4,  5,  6],
         [ 7,  8,  9, 10, 11, 12]],

        [[ 1,  2,  3,  4,  5,  6],
         [ 7,  8,  9, 10, 11, 12]],

        [[ 1,  2,  3,  4,  5,  6],
         [ 7,  8,  9, 10, 11, 12]]])

Stacked tensor shape (dim=-1): torch.Size([2, 6, 3])
Stacked tensor (dim=-1):
 tensor([[[ 1,  1,  1],
         [ 2,  2,  2],
         [ 3,  3,  3],
         [ 4,  4,  4],
         [ 5,  5,  5],
         [ 6,  6,  6]],

        [[ 7,  7,  7],
         [ 8, 

**Arithmetic operations**

Some of the standard math operators are "overloaded", meaning that they work directly with tensors when one of the operands is a tensor.

In [22]:
a = torch.tensor([1, 2, 3])
b = torch.tensor([4, 5, 6])
print("\nAdd tensors:", a + b)

# Element-wise multiplication
print("Multiply tensors:", a * b)

# Dot-product (element-wise multiplication, then sum)
print(torch.dot(a, b))


Add tensors: tensor([5, 7, 9])
Multiply tensors: tensor([ 4, 10, 18])
tensor(32)


But be care if the tensors are not of the same shape. Notice, I'm using try/except to catch the error.

In [23]:
try:
    a = torch.tensor([1, 2, 3])
    b = torch.tensor([4, 5])

    print("\nAdd tensors:", a + b)
except RuntimeError as e:
    print(f"Error adding tensors: {e}")

Error adding tensors: The size of tensor a (3) must match the size of tensor b (2) at non-singleton dimension 0


#### Matrix Multiplication Reference Table

| Multiplication Name | PyTorch Function / Operator | Description |
| :--- | :--- | :--- |
| **Element-wise Multiplication** | `*` or `torch.mul` | Multiplies corresponding elements of two tensors of the same shape (Hadamard product). |
| **Dot Product** | `torch.dot` | Computes the inner product of two 1D tensors (vectors). |
| **Matrix Multiplication** | `@` or `torch.matmul` | Performs standard matrix-matrix multiplication, including broadcasting for batches. |
| **Batch Matrix Multiplication** | `torch.bmm` | Performs matrix multiplication on two batches of matrices without broadcasting. |
| **Matrix-Vector Product** | `torch.mv` | Performs a multiplication of a matrix by a vector. |
| **Outer Product** | `torch.outer` | Computes the outer product of two vectors, resulting in a matrix. |

**Key Concepts:**
*   **Batch**: A collection of multiple samples (like images or data rows) grouped together in a single tensor. In PyTorch, the first dimension is typically the "batch dimension" (e.g., a shape of `[64, 3, 224, 224]` represents a batch of 64 color images).
*   **Matrix Multiplication Rules**: To multiply matrix $A$ (shape $m \times n$) by matrix $B$ (shape $p \times q$), the "inner" dimensions must match ($n = p$). The resulting matrix will have the "outer" dimensions ($m \times q$).
*   **Inner vs. Outer Product**:
    *   **Inner Product (Dot Product)**: Takes two vectors and returns a single scalar (sum of element-wise products).
    *   **Outer Product**: Takes two vectors of size $m$ and $n$ and returns an $m \times n$ matrix where each element $(i, j)$ is the product of the $i$-th element of the first vector and the $j$-th element of the second.
*   **Broadcasting**: A mechanism that allows PyTorch to perform operations on tensors of different shapes by automatically "expanding" the smaller tensor to match the dimensions of the larger one without copying data.


In [24]:
# For matrix multiplication, let's prepare two tensors
v = torch.tensor([5, 10, 15])
a = torch.tensor([[1, 2, 3],
                  [4, 5, 6]])  # shape (2×3)
b = torch.tensor([[7, 8],
                  [9, 10],
                  [11, 12]])  # shape (3×2)

# Demonstrate an inner product (dot product) with v
v_inner = torch.dot(v, v)
print("\nInner product of v with itself:", v_inner)

# Demonstrate a standard multiplication on v
v_mul = v * v
print("\nElement-wise multiplication of v by itself:", v_mul)

# Matrix multiplication: result shape is (2×2)
mat_mult = a @ b
print("Matrix multiplication with the @ operator:\n", mat_mult)

# Or use the torch.matmul function
print("Matrix multiplication using torch.matmul:\n", torch.matmul(a, b))

# A few more examples, using the transform with .T
print("A matrix multiplied by its transpose:")
print(a @ a.T)

# Demonstrate a matrix vector product, which is nothign more than a basic matrix multiplication
# but one matrix is a vector
print("Matrix vector product:")
print(a.mv(v))


Inner product of v with itself: tensor(350)

Element-wise multiplication of v by itself: tensor([ 25, 100, 225])
Matrix multiplication with the @ operator:
 tensor([[ 58,  64],
        [139, 154]])
Matrix multiplication using torch.matmul:
 tensor([[ 58,  64],
        [139, 154]])
A matrix multiplied by its transpose:
tensor([[14, 32],
        [32, 77]])
Matrix vector product:
tensor([ 70, 160])


**Single-element tensors** If you have a one-element tensor, for example
by aggregating all values of a tensor into one value, you can convert it
to a Python numerical value using `item()`. It's a way of "unboxing" the tensor into a Python scalar.


In [25]:
agg = a.sum()
print(f"sum of all elements in:\n{a}\n = {agg}\n(type = {type(agg)})")

print("\nGrabbing the single element and converting to a Python number:")
agg_item = agg.item()
print(agg_item, type(agg_item))

sum of all elements in:
tensor([[1, 2, 3],
        [4, 5, 6]])
 = 21
(type = <class 'torch.Tensor'>)

Grabbing the single element and converting to a Python number:
21 <class 'int'>


**In-place operations** Operations that store the result into the
operand are called in-place. They are denoted by a `_` suffix. For
example: `x.copy_(y)`, `x.t_()`, will change `x`.


In [26]:
print("Original tensor a:\n", a, "\n")

# Standard addition (creates a new tensor, does not modify 'a')
a_added = a + 5
print("Result of a + 5 (not inplace):\n", a_added)
print("Tensor a after standard addition (should be unchanged):\n", a, "\n")

# In-place addition (modifies 'a' directly)
a.add_(5)
print("Tensor a after in-place addition (a.add_(5)):\n", a)

Original tensor a:
 tensor([[1, 2, 3],
        [4, 5, 6]]) 

Result of a + 5 (not inplace):
 tensor([[ 6,  7,  8],
        [ 9, 10, 11]])
Tensor a after standard addition (should be unchanged):
 tensor([[1, 2, 3],
        [4, 5, 6]]) 

Tensor a after in-place addition (a.add_(5)):
 tensor([[ 6,  7,  8],
        [ 9, 10, 11]])


In [27]:
%%html
<div class="note-box">
<h3>Note</h3>
<p>In-place operations save some memory, but can be problematic when computing derivatives because of an immediate loss of history. So, be careful! In most cases, at least in Pytorch, their use is discouraged.</p>
</div>

In [28]:
# Using our tensor from our previous example
print(">\nOriginal tensor a:\n", a, "\n")
# Convert to numpy array
a_numpy = a.numpy()
print(f"Numpy version:\n{a_numpy}")

# Convert to a list
a_list = a.tolist()
print(f"List version:\n{a_list}")

>
Original tensor a:
 tensor([[ 6,  7,  8],
        [ 9, 10, 11]]) 

Numpy version:
[[ 6  7  8]
 [ 9 10 11]]
List version:
[[6, 7, 8], [9, 10, 11]]



NOTE: To convert from a list to a tensor, just create a new tensor with the list as the argument.

In [29]:
# Converting back...
c = torch.from_numpy(a_numpy)
print(c)

c = torch.tensor(a_list)
print(c)

tensor([[ 6,  7,  8],
        [ 9, 10, 11]])
tensor([[ 6,  7,  8],
        [ 9, 10, 11]])


In [30]:
%%html
<div class="todo-box">
<h3>TODO</h3>
</div>

### Question Set 3 - Basic Tensor Operations

1. Create a 3x3 tensor filled with random values sampled from a normal distribution (mean=0, variance=1) using the `torch.randn()` function.


In [31]:
# ANSWER - 1
tensor = torch.randn(3,3)
print(tensor)



tensor([[-1.8882, -0.2616,  0.7786],
        [ 0.0197, -0.1692, -1.1849],
        [ 0.2005,  0.6108, -0.6838]])


2. What is the `...` in the above example? How does it compare to using `:` to select all elements?


**ANSWER:**<br>The `...` is the ellipsis and it is a placeholder for as many colons as needed to select the remaining unspecified dimensions while the `:` is the colon that selects all elements along a single, specific dimension. [source](https://discuss.pytorch.org/t/difference-between-and-in-tensor-slicing/130443)





3. Create a 1D tensor of size 5 with evenly spaced values between 0 and 1 (inclusive). Use the `torch.linspace()` function.

In [32]:
# ANSWER - 3
tensor = torch.linspace(0, 1, 5)
print(tensor)



tensor([0.0000, 0.2500, 0.5000, 0.7500, 1.0000])



**What is a batch?**<br>
The term "batch" is basically a collection of multiple matrices processed together. Thus, batches are usually tensors with more than two dimensions. This is oversimplified, but it's a good starting point.

Think about input data processing. You might have millions of examples that need to be fed into your neural network, each example has perhaps 1000 features. You can't feed the entire dataset all at once, so you batch them. It's common to have batches of 32, 64, 128, 256, 512, 1024, etc. examples. The first dimension of the tensor is the batch dimension. This is particularly useful for training neural networks on GPUs, where you can process multiple examples in parallel. It's also useful when dealing with color images, where each pixel is a feature over 3 values (red, green, blue). So a batch of 32 images is a 32x32x3 tensor.


For example, if you have a batch of 5 matrices, each 4x4, you can create a tensor with shape (5, 4, 4).
```python
batch = torch.randn(5, 4, 4)
```
This creates a 3D tensor with shape (5, 4, 4), which means:
* 5 matrices (the batch dimension)
* Each matrix is 4×4<br>

So you have 5 4x4 matrices stacked together. The first dimension (5) is the batch dimension.

**Why use batches?**<br>
* Generally, it comes down to efficiency: Process multiple examples at once using parallel operations. It's quite common in neural networks, especially training ones. Neural networks typically process data in batches rather than one example at a time
* Vectorization: PyTorch can perform operations on all matrices in the batch simultaneously

Back to the questions...

4. Generate a **batch** of 5 randomly initialized 4x4 matrices whose values are drawn from a uniform distrution between 10 and 100. Then, compute the matrix-vector product with a random vector of 4 values whose values are drown from a normal distribution with mean 0 and standard deviation 1. What is the shape of the resulting tensor? Show the computation and output the result and its shape. (The result should be a 5x4 tensor.)

WARNING: Be careful of the types.

In [33]:
# ANSWER - 4
batch = (100 - 10) * torch.rand(5, 4, 4) + 10
print(f"Original: {batch}\n{batch.shape}")

vec = torch.randn(4)

processed = batch @ vec

print(f"Processed: {processed}\n{processed.shape}")


Original: tensor([[[69.8472, 70.9613, 27.8086, 12.1084],
         [15.0969, 59.5141, 89.6376, 36.1417],
         [48.1531, 68.4018, 60.0465, 30.2142],
         [60.9319, 97.0879, 82.0451, 56.7112]],

        [[77.8027, 66.6565, 52.4113, 18.3973],
         [28.5887, 69.1697, 40.5298, 18.5508],
         [18.1046, 28.0019, 39.0296, 78.8107],
         [27.5942, 20.6050, 61.3403, 24.2666]],

        [[72.2820, 91.2350, 94.4291, 14.6170],
         [76.9962, 58.1218, 25.9168, 85.5970],
         [57.7320, 33.9925, 60.9631, 63.2950],
         [64.2506, 32.1253, 80.0012, 28.1934]],

        [[67.5129, 38.9049, 26.7707, 14.6015],
         [31.9285, 29.9077, 32.3800, 88.4898],
         [49.6521, 14.6101, 48.5936, 32.9898],
         [10.0923, 56.5840, 46.9468, 32.8478]],

        [[52.9433, 54.0243, 27.9309, 44.9824],
         [63.6212, 69.8339, 67.6186, 30.7487],
         [51.2583, 30.7157, 12.7231, 84.8009],
         [77.3385, 16.4975, 21.4807, 24.1753]]])
torch.Size([5, 4, 4])
Processed: tensor(

**What is broadcasting?**<br>

Broadcasting is a technique used by PyTorch (and other array libraries like NumPy) to perform operations on tensors of different shapes. When operating on two tensors, PyTorch will automatically expand the smaller tensor's shape so that it matches the larger tensor's shape in a compatible way, without actually copying data. This is very useful for element-wise operations like addition, subtraction, etc.

**Rules of Broadcasting:**
- Starting from the trailing dimensions, two dimensions are compatible when they are equal, or one of them is 1.
- If necessary, the dimensions of size 1 are stretched or "broadcast" to match the other tensor.


**Example:**<br>
Create a 3x4 tensor filled with the value 2 named 'a' and a 1D tensor of size 4 with values [1, 2, 3, 4] named 'b'. Add the two tensors together. Print the result and explain the shape of the output.


In [34]:
a = torch.full((3, 4), 2)
print("Tensor a:\n", a)
#
# Create a 1D tensor of size 4
b = torch.tensor([1, 2, 3, 4])
print("Tensor b:\n", b)
#
# Add using broadcasting
# 1. Starting from the last dimension, b (size 4) matches a (size 4).
# 2. b has no first dimension, so PyTorch treats it as having size 1.
# 3. This size 1 is "broadcast" (virtually repeated) 3 times to match a.
# 4. Effectively, [1, 2, 3, 4] is added to each of the 3 rows in 'a'.
result = a + b
print("Result of a + b:\n", result)
print("Result shape:", result.shape)

Tensor a:
 tensor([[2, 2, 2, 2],
        [2, 2, 2, 2],
        [2, 2, 2, 2]])
Tensor b:
 tensor([1, 2, 3, 4])
Result of a + b:
 tensor([[3, 4, 5, 6],
        [3, 4, 5, 6],
        [3, 4, 5, 6]])
Result shape: torch.Size([3, 4])


Back to the questions...

5. Create a 5x5 tensor of zeros named 't'. Set the diagonal elements to 1. Add 5 to every element in the second row. Print the result and the shape of the output.


In [56]:
# ANSWER - 5
t = torch.zeros((5,5))
t.fill_diagonal_(1)
t[1,:] += 5

print(t)
print(t.shape)




tensor([[1., 0., 0., 0., 0.],
        [5., 6., 5., 5., 5.],
        [0., 0., 1., 0., 0.],
        [0., 0., 0., 1., 0.],
        [0., 0., 0., 0., 1.]])
torch.Size([5, 5])


### 1.4 `torch.distributions`: a quick overview

`torch.distributions` provides **probability distributions** you can sample from and evaluate. Each distribution is a Python object (e.g., `Normal`, `Bernoulli`, `Exponential`) with methods like:
- `.sample(shape)` or `.rsample(shape)` to draw random samples,
- `.log_prob(x)` to compute log-likelihoods,
- `.mean`, `.variance`, and other properties when available.

The most common distributions you'll see in machine learning are the **Normal** distribution and the **Uniform**:
- **Normal(μ, σ)**: symmetric bell-shaped distribution for continuous data. **This is, by far, the most common distribution in machine learning.**
- **Uniform(a, b)**: continuous values equally likely between `a` and `b`.

There are many other distributions you may come across in practice, depending on the problem at hand:
- **Bernoulli(p)**: binary outcomes (0 or 1) with success probability `p`.
- **Binomial(n, p)**: number of successes in `n` trials with success probability `p`.
- **Categorical(probs)**: one of K discrete classes.
- **Exponential(λ)**: continuous waiting-time distribution with rate `λ`.

For example, you might use a **Bernoulli** distribution to model coin-flip outcomes, a **Binomial** to model n independent coin-flips, or a **Categorical** distribution to sample class labels in a classifier.

We’ll use the **Exponential** distribution in the challenge below.

In [36]:
%%html
<div class="todo-box">
<h3>TODO</h3>
</div>

### Question Set 4 - Distributions

1. What is the difference between `.sample()` and `.rsample()` in `torch.distributions`?

**ANSWER:**<br> `.sample()` is a random sampling from a probability distribution. This means that they are simply values that are sampled from the distribution, so they are detached from gradients. `.rsample()`, on the other hand, returns samples that are reparametrized so that the gradients are still attached to the samples. This is done by separating the randomness from parameters that the gradient can flow through. [source](https://stackoverflow.com/questions/60533150/what-is-the-difference-between-sample-and-rsample)







2. In your own words, what is a **log likelihood**, and why do we often use the log?

**ANSWER:**<br>The log likelihood is just the natural logarithm of the probability/likelihood of an event occuring. We use log because we can use the product rule property where the log of products is equal to the sum of the individual logs. It is also an increasing function, so it preserves ordinality. [source](https://www.reddit.com/r/AskStatistics/comments/l14whl/can_someone_explain_log_likelihood_in_an/)






3. What does `.log_prob(x)` return, and how is it related to the likelihood of `x`?

**ANSWER:**<br>`.log_prob(x)` returns a tensor of the natural logarithm of the likelihood of `x` in the given distribution. 





4. Create a tensor of 10,000 samples from an **Exponential** distribution with rate $\lambda = 2$.
Then compute and print:
* The shape of the tensor.
* The empirical mean and variance.
* The fraction of samples greater than $1.0$.

_Hint:_ Theoretical mean is $\frac{1}{\lambda}$ and variance is $\frac{1}{\lambda^2}$.


In [37]:
# ANSWER - 4

dist = torch.distributions.Exponential(2.0)
tensor = dist.sample((10000,))

print(tensor.shape)
print(tensor.mean())
print(tensor.var())
print((tensor > 1.0).float().mean())






torch.Size([10000])
tensor(0.5000)
tensor(0.2413)
tensor(0.1410)


---

## Section 2: Hardware Acceleration

One of PyTorch’s biggest advantages is that the *same tensor code* can run on different kinds of hardware. By default, your tensors live on the **CPU**, but you can also run them on an **accelerator** (usually a GPU), which can be much faster for large, highly-parallel workloads (like deep neural networks).

PyTorch supports multiple accelerator backends:
- **CUDA** (NVIDIA GPUs — common on Windows/Linux and in Colab)
- **MPS** (Apple Metal Performance Shaders — Apple Silicon Macs)
- Others (e.g., XPU/MTIA), depending on your system and installation

> **What are XPU and MTIA?**  
> - **XPU**: A backend name used in PyTorch to refer to certain **Intel accelerators**, most commonly **Intel GPUs** (e.g., via Intel’s oneAPI/Level Zero stack and the Intel Extension for PyTorch). If your PyTorch build includes this support, you may see devices like `"xpu"` in addition to `"cpu"`/`"cuda"`/`"mps"`.  
> - **MTIA**: **Meta Training and Inference Accelerator**, a **custom AI accelerator** developed by Meta for internal training/inference workloads. Some PyTorch builds/environments may expose `"mtia"` as a device type, but it’s typically only available in Meta’s infrastructure.

To learn more, see the PyTorch docs: [Accelerators](https://pytorch.org/docs/stable/torch.html#accelerators)

### Getting an accelerator

- **Google Colab:** `Runtime > Change runtime type > GPU`
- **Windows (local) / Linux (local):** requires an NVIDIA GPU *and* a CUDA-enabled PyTorch install (plus the CUDA toolkit/driver setup)
- **macOS (local):** MPS is available on **Apple Silicon** (M1/M2/M3/…) and is not available on Intel Macs. Fortunately, Apple has been contributing to the PyTorch ecosystem and the standard PyTorch installation now includes MPS support.

### Default behavior
By default, tensors are created on the CPU:


In [38]:
t = torch.rand(10000,1000)

# Let's show the device the tensor is running on
print(t.device)

cpu


In [39]:
%%html
<div class="note-box">
<h3>NOTE</h3>
<p>t.device is not a string type! It is an object of type torch.device! You are just getting the repr string above.</p>
</div>

We need to explicitly move
tensors to the accelerator using the `.to` method (after checking for
accelerator availability). Keep in mind that copying large tensors
across devices can be expensive in terms of time and memory, and it doesn't always make sense to do so for simple networks and small datasets. More on that shortly.

What device is available for you? Well, it depends on your system. You can evaluate your OS using the `sys` package:

* Windows: 'win32'
* MacOS: 'darwin'
* Linux: 'linux'


In [40]:
import sys
print(f"Operating System: {sys.platform}")

if sys.platform == 'darwin':
  print("Apple Silicon")
elif sys.platform == 'linux':
  print("Linux")
elif sys.platform == 'win32':
  print("Windows")
else:
  print("Unknown")

Operating System: linux
Linux


Generally, while there are plenty of cases where it doesn't make sense to use your GPU, for large or deep neural networks or when dealing with a large, high-dimensional dataset for training, you will want to use your GPU as your backend.

### Checking for accelerator availability

PyTorch generally exposes backend availability of GPU-based backends depending on the system.

* Windows/Linux: `torch.cuda.is_available()` - this will check to see if you have the CUDA toolkit installed and available on your system
* Mac: `torch.backends.mps.is_available()` - **MPS** stands for **Metal Performance Shaders**. It's Apple's framework designed for high-performance GPU-accelerated computation, specifically for Apple Silicon Macs (like those with M1, M2 chips, etc.). It allows machine learning frameworks like PyTorch to utilize the integrated GPU on these Apple devices for accelerated tasks.

In the early days of Pytorch, you would have to check for specific backends relevant to your environment or desired hardware.

For example:

In [41]:
mps_available = torch.backends.mps.is_available()
mps_built = torch.backends.mps.is_built()
print(f"MPS available: {mps_available}")
print(f"MPS built: {mps_built}")

MPS available: False
MPS built: False


In [42]:
print(f"CUDA available: {torch.cuda.is_available()}")

CUDA available: True


So, we could store specific device availability as some booleans:

In [43]:
# Check for CUDA (Linux/Windows with NVIDIA GPU)
cuda_available = torch.cuda.is_available()
# Check for MPS (Apple Silicon Macs)
mps_available = torch.backends.mps.is_available()

device = "cuda" if cuda_available else "mps" if mps_available else "cpu"
print(f"Accelerator device: {device}")

Accelerator device: cuda


### Best practice - use `torch.accelerator` for device selection

Now, the preferred method is to now use `torch.accelerator` to get the preferred accelerator on your system (if any). You no longer have to have custom code that checks for specific backends:

In [44]:
# True if *some* accelerator is available (CUDA, MPS, etc.)
has_accel = torch.accelerator.is_available()
print(f"Accelerator available: {has_accel}")

# A torch.device representing the preferred accelerator (if any). If none available, set it to "cpu"
# e.g. torch.device("cuda", 0) or torch.device("mps", 0)
if has_accel:
    accel_dev = torch.accelerator.current_accelerator()
    print(f"Preferred accelerator: {accel_dev}")
else:
    accel_dev = torch.device("cpu")
    print("No accelerator available, defaulting to CPU.")

# Print type of accel_dev to illustrate that it is NOT a string
print(f"Type of accel_dev: {type(accel_dev)}")

Accelerator available: True
Preferred accelerator: cuda
Type of accel_dev: <class 'torch.device'>


Be sure to note above the type of accel_dev. It is an object of type `torch.device`.

### Moving Tensors to Device

By default, tensors are created on the CPU. We need to explicitly move them to the GPU using the `.to()` method.


In [45]:
# Specify at instantiation
t = torch.rand(10000,1000,device=accel_dev)
t.device

device(type='cuda', index=0)

In [46]:
# Or move after you instantiate it:
t = torch.rand(10000,1000)
print(f"Device before: {t.device}")
if has_accel:
    t = t.to(accel_dev)
    print(f"Device after: {t.device}")
else:
    print("No accelerator available, defaulting to CPU.")

Device before: cpu
Device after: cuda:0


### Is a GPU always better?

OK, so this is the question you should be asking. And, well, it depends! GPUs (Graphics Processing Units) are designed for highly parallel computations, meaning they can perform many calculations simultaneously. CPUs (Central Processing Units), while very powerful, are optimized for sequential tasks and complex logic.

So, one might think you should always use a GPU, right? It's not that simple.

#### GPUs explained

GPUs are all about having the power to solve computations with **massive parallelism**! GPUs have thousands of smaller, specialized cores, unlike CPUs which have fewer, but more powerful cores. This architecture allows them to process many pieces of data concurrently.

#### Where do GPUs shine?

GPU cores are highly optimized for mathematical operations, especially matrix multiplications and linear algebra computations, including basic matrix additions and multiplications. Functions that can be parallelized shine quite well on GPUs. These operations are fundamental to fields like graphics rendering, scientific simulations, and yes, **neural networks**, especially large, deep nets.

#### Do GPU computations have a price?

There is a big **Data Transfer Overhead**! For the GPU to perform computations, data first needs to be explicitly moved from the CPU's main memory to the GPU's dedicated high-speed memory (VRAM). After computation, the results often need to be transferred back. This data transfer introduces an inherent overhead.

#### When does a GPU Outperform a CPU?

Generally, you should seek GPU support when:

* **Workload is Large and Highly Parallel**: The task involves a huge number of independent, similar calculations that can be performed at the same time. Examples include processing large images, training deep neural networks, or running complex simulations. The larger the dataset and the more independent calculations involved, the more the GPU's parallel architecture shines.

* **Computation Time Significantly Exceeds Data Transfer Overhead**: The time saved by parallel execution on the GPU must be much greater than the time spent moving data to and from the GPU's memory. For smaller tasks, or tasks that require frequent data transfers back and forth between the CPU and GPU, the overhead can negate or even reverse the GPU's speed advantage.

* **High Throughput Required**: When the goal is to process a massive amount of data or perform a large number of computations in the shortest possible time, GPUs excel at achieving high throughput.

In summary, while CPUs are versatile and efficient for a wide range of tasks, GPUs become indispensable for workloads that are (to quote one thread I read), 'embarrassingly parallel' and involve heavy numerical computation on large datasets. **The more intense and parallelizable the computation, the greater the performance gap between a GPU and a CPU will be.**

### Demo code

Let's try a  demanding workload with significantly large tensors to truly see the GPU's advantage. You can tweak the tensor sizes for matrix multiplication and convolution to see the impact.

In [47]:
import torch
import time
import platform

def run_massive_tensor_computations(size: int, device: torch.device = torch.device("cpu")):
    print(f"\nRunning massive computations on device: {device}")
    torch.manual_seed(42)

    # Generate much larger random tensors
    size_matmul = (size, size)  # Doubled from previous example
    tensor_a = torch.randn(size_matmul, device=device)
    tensor_b = torch.randn(size_matmul, device=device)

    # For convolution, increase dimensions
    input_tensor_conv = torch.randn(64, 256, 128, 128, device=device) # Larger batch, channels, and spatial dims
    kernel = torch.randn(512, 256, 3, 3, device=device)

    start_time = time.time()

    # --- Complex Computations ---
    # 1. Matrix Multiplication (matmul)
    result_matmul = torch.matmul(tensor_a, tensor_b)

    # 2. Convolution
    result_conv = torch.nn.functional.conv2d(input_tensor_conv, kernel, padding=1)

    # 3. Element-wise operations (addition, multiplication) - still relatively fast
    result_elem_ops = (result_matmul * 0.5 + 0.1) / result_matmul.mean()

    # 4. Activation function (ReLU)
    result_relu = torch.relu(result_elem_ops)

    # 5. Further matrix multiplication with a smaller tensor (still large)
    smaller_tensor = torch.randn(size_matmul[0], size//4, device=device)
    result_final = torch.matmul(result_relu[:, :size_matmul[0]], smaller_tensor)

    if device.type != "cpu":
        # Wait for GPU to finish computations
        torch.accelerator.synchronize()

    end_time = time.time()
    print(f"Total computation time: {end_time - start_time:.4f} seconds")

# Run our tests!
size=16384
# size=1024

print(f"Testing on CPU {platform.processor}:")
run_massive_tensor_computations(size=size,device=torch.device("cpu"))

if has_accel:
    print(f"Testing on {accel_dev}:")
    run_massive_tensor_computations(size=size,device=accel_dev)
else:
    print("No accelerator available. WHOMP WHOMP...")

Testing on CPU <function processor at 0x7f11ee52a520>:

Running massive computations on device: cpu
Total computation time: 19.2744 seconds
Testing on cuda:

Running massive computations on device: cuda
Total computation time: 1.0721 seconds


Preliminary tests on above code

With size=16384:

Google Colab:
* cpu = 19.84 sec
* cuda = 1.03 sec (using A100 w/ high ram)
* cuda = 1.6012 sec (using L4 GPU)

On Macbook Pro w/ M2 Max (2023) Processor:
* cpu = 11.0712 sec
* mps = 1.4997 sec

In [96]:
%%html
<div class="todo-box">
<h3>TODO</h3>
</div>

### Question Set 5 - Accelerator devices

1. Why don't we just move everything to the GPU immediately?

**Answer:**<br> Moving everything to the GPU immediately can be inefficient for smaller and less computational-heavy workloads as the I/O transfer time will dominate it, making it inefficient.





2. Modify the code above to run the above on a size=1024 and size=16384, and record your results. If you don't have a GPU on your local system, then either use Colab or just report the results you get on your cpu.

**ANSWER:**<br>

Report your results here!

<b>size=1024</b><br>
Testing on CPU:

Running massive computations on device: cpu<br>
Total computation time: 4.5689 seconds<br>

Testing on cuda:

Running massive computations on device: cuda<br>
Total computation time: 0.0252 seconds<br>

<b>size=16384</b><br>
Testing on CPU <function processor at 0x7f11ee52a520>:

Running massive computations on device: cpu<br>
Total computation time: 19.2744 seconds<br>

Testing on cuda:

Running massive computations on device: cuda<br>
Total computation time: 1.0721 seconds<br>

---
## Section 3: Automatic Differentiation (Autograd)

`torch.autograd` is PyTorch’s built-in differentiation engine that powers neural network training. It automatically computes gradients for any computational graph. You will soon appreciate the important of highly efficient differentiation!

### What is Automatic Differentiation?

**Automatic differentiation (autodiff)** is a technique for efficiently computing derivatives of functions specified by computer programs. Unlike symbolic differentiation (which manipulates mathematical expressions) or numerical differentiation (which uses finite differences), autodiff computes exact derivatives while maintaining computational efficiency.

In the context of neural networks, we need to compute gradients of a loss function with respect to thousands or even millions of parameters. Doing this by hand would be intractable. Automatic differentiation solves this problem elegantly.

### Why is Autograd Critical for Deep Learning?

Neural network training relies on an algorithm called **backpropagation**, which is essentially the chain rule of calculus applied repeatedly. Here's why autograd is essential:

1. **Gradient-Based Optimization**: Training a neural network means finding parameters (weights and biases) that minimize a loss function. This requires computing gradients of the loss with respect to every parameter.

2. **Complex Computational Graphs**: Modern neural networks involve complex operations—matrix multiplications, non-linear activations, convolutions, attention mechanisms, etc. Computing gradients through these operations manually is error-prone and tedious.

3. **Efficiency**: Autograd computes gradients in a single backward pass through the computational graph, making it highly efficient even for very deep networks.

4. **Flexibility**: PyTorch's dynamic computational graph (define-by-run) means you can change your network architecture on the fly, and autograd will still work correctly.


### How Does PyTorch's Autograd Work?

PyTorch uses a technique called **reverse-mode automatic differentiation**. Here's the key idea:

#### 1. Building the Computational Graph

When you perform operations on tensors with `requires_grad=True`, PyTorch builds a **directed acyclic graph (DAG)** called the computational graph:
- **Nodes** represent tensors (data)
- **Edges** represent operations (functions)

Each operation creates a `grad_fn` object that knows how to compute gradients!

In [48]:
import torch
# Example: Building a simple computational graph
x = torch.tensor(3.0, requires_grad=True)
y = torch.tensor(4.0, requires_grad=True)

# Operations build the graph
z = x * y + y ** 2
print(f"z = {z}")
print(f"z.grad_fn = {z.grad_fn}")  # Shows the operation that created z

z = 28.0
z.grad_fn = <AddBackward0 object at 0x7f1090260ca0>


In this example:
- `x` and `y` are **leaf tensors** (created by the user, not by operations)
- `z` is a **non-leaf tensor** created by operations
- `z.grad_fn` is `<AddBackward0>`, representing the addition operation

#### 2. The Backward Pass

When you call `.backward()` on a scalar tensor (like a loss), PyTorch:
1. Starts at that tensor
2. Traverses the graph backward (from outputs to inputs)
3. Applies the chain rule at each node to compute gradients
4. Accumulates gradients in the `.grad` attribute of each leaf tensor

This is going to make so much more sense when we work through backpropagation - is the process by which neural nets computes gradients for all tensors in the computational graph, starting from the loss and moving backward through the operations that created them. This allows us to efficiently compute gradients for complex models with many interconnected operations.

Think about this mathematically for a second. Our basic function, z, is:

$$z = x \cdot y + y^2$$

The derivative of z with respect to x is y, and with respect to y is x + 2y. Let's verify this, but first we need to explain the `.backward()` method:

#### Understanding `z.backward()`

When we call `z.backward()`, PyTorch performs the **backward pass** through the computational graph:

1. **Starting Point**: It begins at tensor `z` (our output) and works backward toward the inputs `x` and `y`.

2. **Chain Rule Application**: At each operation node in the graph, PyTorch applies the chain rule to compute how much each input contributed to the final value. For our function $z = x \cdot y + y^2$:
   - The gradient $\frac{\partial z}{\partial x} = y$ (derivative of $xy$ with respect to $x$)
   - The gradient $\frac{\partial z}{\partial y} = x + 2y$ (sum of derivatives: $\frac{\partial(xy)}{\partial y} + \frac{\partial(y^2)}{\partial y}$)

3. **Gradient Storage**: The computed gradients are stored in the `.grad` attribute of each **leaf tensor** (tensors created by the user with `requires_grad=True`). In our case, `x.grad` and `y.grad` will contain the respective partial derivatives.

4. **Graph Destruction**: After `.backward()` completes, the computational graph is destroyed by default to free memory (unless `retain_graph=True` is specified).

This automatic differentiation is what makes training neural networks feasible—PyTorch handles all the calculus for us!

OK, let's verify this... and be amazed at the simplicity of how Pytorch carries all of this out for us:

In [ ]:
# Continuing from above

# z.backward 
z.backward()

# Let's verify mathematically:
# z = x*y + y^2
# dz/dx = y = 4.0
# dz/dy = x + 2y = 3 + 2*4 = 11.0

print(f"dz/dx = {x.grad}")  # Should be y = 4.0
print(f"dz/dy = {y.grad}")  # Should be x + 2y = 3 + 2*4 = 11.0

# Source: https://medium.com/@namanphy/understanding-autograd-5-pytorch-tensor-functions-8f47c27dc38

dz/dx = 4.0
dz/dy = 11.0


### Key Concepts in PyTorch's Autograd

1. **Leaf Tensors**: Tensors created directly by the user. Only leaf tensors with `requires_grad=True` will have gradients stored in `.grad`.

2. **Non-Leaf Tensors**: Tensors created by operations. Their gradients are not stored by default (to save memory), but they have `grad_fn` to track how they were created.

3. **Gradient Accumulation**: Calling `.backward()` multiple times will **accumulate** gradients in `.grad`. You need to call `.zero_grad()` to clear them.

In [50]:
# Demonstrating gradient accumulation
x = torch.tensor(2.0, requires_grad=True)

for i in range(3):
    y = x ** 2 + 2 * x
    # dy/dx = 2x + 2
    y.backward()
    print(f"Iteration {i + 1}: x.grad = {x.grad}")

# Notice gradients accumulate! To prevent this:
x.grad.zero_()
print(f"After zeroing: x.grad = {x.grad}")

Iteration 1: x.grad = 6.0
Iteration 2: x.grad = 12.0
Iteration 3: x.grad = 18.0
After zeroing: x.grad = 0.0


4. **Detaching from the Graph**: Sometimes you want to use a tensor's value without tracking gradients. Use `.detach()` to create a new tensor that shares data but has no gradient tracking.

In [51]:
x = torch.tensor(3.0, requires_grad=True)
y = x ** 2

# Detach y from the computational graph
y_detached = y.detach()
print(f"y_detached requires_grad: {y_detached.requires_grad}")

z = y_detached * 2

# Now z has no connection to x, so gradients won't flow back
# z.backward() would fail because z is not connected to any requires_grad=True tensor

y_detached requires_grad: False


### Memory Management

PyTorch automatically manages the computational graph:
- The graph is built dynamically during the forward pass
- After `.backward()`, the graph is destroyed by default (to save memory)
- For multiple backward passes, use `retain_graph=True` in `.backward()`

### Practical Implications

Understanding autograd is crucial because:
1. **Debugging**: You need to understand when gradients flow and when they don't
2. **Custom Operations**: You may need to define custom backward passes for novel operations
3. **Optimization**: Knowing when to detach or use `torch.no_grad()` can save memory and computation
4. **Advanced Architectures**: Some architectures (like GANs) require careful gradient management

Don't get too concerned about the details right now. Just know that the automatic differential is absolutely critical for neural network training.

---
**End of Lab 01**


In [52]:
%%html
<div class="section-box">
<h1></h1>
</div>

# <div class="section-box">Challenges</div>

The **Challenges** section is where things get more advanced... and that’s the point. These problems are designed as hands-on, active learning challenges that push you past “follow-the-steps” and into real understanding. For this first assignment, you'll be applying tensors, broadcasting, devices, and autograd in new situations. You already have everything you need to succeed: the lab examples, the PyTorch docs, your reasoning skills, and the willingness to experiment and debug. And yes, you can use AI to assist you! Just keep track of how you're using it. Don't ask for answers. Ask for guidance so you can get the answers. Expect a little struggle! That’s your brain growing. If you don't struggle a bit and take time to think through, you're not learning.

You’ve got this. These are easy questions that will help you get comfortable with the material. Look for much more challenging questions in coming assignments!


## 1. Broadcasting & Batching
Suppose you have a batch of 10 images. Each image is 32x32 pixels. Each pixel has 3 color channels (RGB). You want to normalize each channel across the entire batch by subtracting a specific mean value for each channel: Red: 0.485, Green: 0.456, Blue: 0.406.
   * Create a random tensor representing this batch if images (i.e your shape should be: `[10, 3, 32, 32]`).
   * Create a mean_adj tensor.
   * Subtract means_adj from the batch (HINT: Use broadcasting, and align the dimensions correctly!)
   * Show the 3-channel means before and after, and verify the shape of the result.


In [58]:
# Answer 1

img_batch = torch.rand((10, 3, 32, 32))

mean_adj = torch.tensor([0.485, 0.456, 0.406])

mean_adj = mean_adj.reshape(3, 1, 1)

new_batch = img_batch - mean_adj

print(f"Before: {img_batch}")

print(f"After: {new_batch}")







Before: tensor([[[[1.3043e-01, 2.2600e-01, 1.6154e-01,  ..., 1.8219e-01,
           4.8849e-01, 4.2874e-01],
          [2.3789e-01, 3.7786e-01, 1.4952e-01,  ..., 7.5504e-01,
           5.8019e-01, 6.5616e-01],
          [9.1765e-01, 4.5308e-01, 7.3544e-01,  ..., 5.8250e-02,
           3.7011e-01, 5.4518e-01],
          ...,
          [3.8926e-01, 3.8063e-01, 5.5742e-01,  ..., 3.6854e-01,
           1.9082e-02, 4.9861e-01],
          [9.8492e-01, 1.2661e-01, 9.5162e-01,  ..., 8.7402e-01,
           4.3856e-01, 6.1520e-01],
          [4.2341e-01, 4.6359e-01, 5.4169e-02,  ..., 1.0876e-01,
           3.6204e-01, 8.6206e-01]],

         [[5.4352e-01, 6.0351e-01, 5.2461e-02,  ..., 2.7743e-01,
           2.1191e-01, 4.7134e-01],
          [1.4189e-01, 6.1758e-01, 9.0812e-01,  ..., 4.2118e-01,
           9.8032e-01, 8.7665e-01],
          [5.6920e-01, 4.7605e-01, 9.0159e-01,  ..., 8.2211e-01,
           7.6823e-02, 5.7308e-01],
          ...,
          [5.8744e-01, 8.0014e-01, 6.6077e-01,  ...

## 2. Autograd & Inference Mode
You will soon see that in training loops, we use `optimizer.zero_grad()` to clear gradients. However, when performing evaluation (inference) with a trained model, we want to disable gradient tracking entirely to save memory and compute resources.
   * Create a tensor `w` with `requires_grad=True` representing a vector of 3 weights (e.g., `[2.0, 3.0, 4.0]`)
   * Perform a calculation $z = w^2 + 5$ inside a `torch.no_grad()` Python block
   * Then, perform the same calculation using `w.detach()`
   * Check if the result has a `grad_fn` attribute (they should both be `None`).
   * As a comment at the end, explain the difference between using `torch.no_grad()` and `.detach()`.


In [ ]:
# Answer 2
w = torch.tensor([2.0, 3.0, 4.0], requires_grad=True)

with torch.no_grad():
    z = w ** 2 + 5

print(f"Using no_grad context manager: {z.grad_fn}")

w = w.detach()
z = w ** 2 + 5

print(f"Using .detach(): {z.grad_fn}")

# Explanation
# torch.no_grad() is a context manager, which means that any code inside the block is in an environment/context where gradients are not tracked. In this case,
# z is a tensor that is created with its gradients not being tracked. .detach(), on the other hand, creates a whole new tensor which has no gradient history and 
# does not have its gradient being tracked.



Using no_grad context manager: None
Using .detach(): None


## 3. `reshape` vs. `view` and the impact on contiguity
Explain *when* `tensor.view(...)` will fail but `tensor.reshape(...)` will succeed. <br><br>Then **demonstrate this with code**. Using a **transpose** to create a non-contiguous tensor.

You can start by initializing `t` with this tensor:<br>
`t = torch.arange(16).reshape(4, 4)` (a 4×4 tensor containing 0..15).

Then, work through an example trying to view the transpose of `t` as a single vector of 16 elements. If you do it right, it should look a follows:<br>
`tensor([ 0,  4,  8, 12,  1,  5,  9, 13,  2,  6, 10, 14,  3,  7, 11, 15])`

**What your answer must include:**
1. A short explanation (1–3 sentences) of why contiguity matters for `.view()` as a comment at the top of your code cell
2. Python code that:
   - Creates `t` exactly as specified above.
   - Creates a transposed tensor
   - Prints whether `t` and the transposed tensor use the same memory location and are contiguous.
   - Tries `view` on the transposed tensor to show the data as a single vector of 16 elements (catch the `RuntimeError`).
   - Uses `reshape` on the transposed tensor to flatten it (e.g., to shape `(16,)`) and prints the resulting shape.

**Hints (methods you’ll likely use):**
- `.T` or `.transpose(...)`
- `.is_contiguous()`
- `.view(...)` and `.reshape(...)`
- Optional: `.contiguous()` (to show how to “fix” the tensor so `.view()` works)


In [ ]:
# Answer 3
# Contiguity is required for .view() because it uses the same underlying data (as a View) while .reshape() can copy the data if needed (if not contiguous).

t = torch.arange(16).reshape((4, 4))

t_transposed = t.T

print(f"t: {t.data_ptr()}, {t.is_contiguous()}")
print(f"t_transposed: {t_transposed.data_ptr()}, {t_transposed.is_contiguous()}")

try:
    new_t_transposed = t_transposed.view((16,))
except RuntimeError:
    print("Cannot view the t_transposed as a single vector of 16 elements")

new_t_transposed = t_transposed.reshape((16,))

print(f"new_t_transposed: {new_t_transposed}, {new_t_transposed.shape}")









t: 1228762880, True
t_transposed: 1228762880, False
Cannot view the t_transposed as a single vector of 16 elements
new_t_transposed: tensor([ 0,  4,  8, 12,  1,  5,  9, 13,  2,  6, 10, 14,  3,  7, 11, 15]), torch.Size([16])


In [62]:
%%html
<div class="section-box">
<h1></h1>
</div>

# <div class="section-box">Career Surveys</div>


Let's turn you into a "Market Researcher" using AI to map out what's going on in this space.

**Objective:** Use Generative AI (Gemini, Claude, or ChatGPT) to identify the current "hot spots" in the AI job market and translate industry jargon into actionable skills.


In [63]:
%%html
<div class="todo-box">
<h3>TODO</h3>
</div>

## Part 1: The AI Scout

Looking for a job? For Pete's sake, don't just search LinkedIn. (And, who's Pete?) Use your favorite LLM to find out where the "real" builders are hanging out.

**Task:** Use an AI to answer the following. Paste your prompts and the AI’s summary below. I'll give you prompts to get you started, but tweak and refine to get the best results. 

1. **Platform Discovery:** *"I am a CS student at Bucknell specializing in PyTorch and MLOps. Beyond LinkedIn, identify the top 3 specialized job platforms for technical AI roles in 2026. Why are these better for engineers than general boards?"*
2. **The Title Decode:** *"Identify the 5 fastest-growing job titles in AI for 2026. For each title, identify if they typically require 'Training from scratch' (PyTorch/TF) or 'API integration' (HuggingFace/OpenAI)."*

Be sure to record what AI engine you used.

#### ANSWER (Using Claude):

1. **Platform Discovery:** *"I am a CS student at Bucknell specializing in PyTorch and MLOps. Beyond LinkedIn, identify the top 3 specialized job platforms for technical AI roles in 2026. Why are these better for engineers than general boards?"*<br>

```
Based on current trends, here are the top 3 specialized platforms for AI/ML engineering roles:
1. Wellfound (formerly AngelList Talent)

Direct access to startup founders and hiring managers
Heavy concentration of early-stage AI companies (Series A-C) building ML infrastructure
Shows equity compensation upfront, tech stack details, and team size
Better than general boards: filters by specific ML frameworks (PyTorch, TensorFlow), infrastructure tools (Kubernetes, Ray), and role type (MLOps, ML infrastructure vs research)

2. AI Jobs Board (ai-jobs.net) / Hugging Face Jobs

Curated specifically for ML/AI positions from companies actively building AI products
Hugging Face board attracts companies serious about open-source ML tooling
Technical depth in job descriptions (model architectures, deployment patterns, scale requirements)
Better than general boards: signal quality—if they're posting here, ML/AI is core to their business, not a side initiative

3. YC Work at a Startup

Access to YC-backed companies (many AI-focused in recent batches)
Direct founder contact, fast interview processes
Strong concentration of infrastructure/tooling companies (not just application layer)
Better than general boards: batch focus areas reveal emerging trends; companies often more willing to sponsor visas for specialized ML talent

Why these beat general boards for your profile:

They filter for companies where PyTorch/MLOps expertise is essential, not nice-to-have
Job descriptions include actual technical depth (infrastructure, serving, training pipeline details)
Higher concentration of roles matching your Meta production engineering + ML infrastructure background
Less competition from general SWE applicants who add "interested in AI" to their resume

Given your KServe contributions and Stellark work, these platforms will surface companies building the infrastructure layer, not just fine-tuning models.
```

2. **The Title Decode:** *"Identify the 5 fastest-growing job titles in AI for 2026. For each title, identify if they typically require 'Training from scratch' (PyTorch/TF) or 'API integration' (HuggingFace/OpenAI)."*

```
5 Fastest-Growing AI Job Titles (2026)
1. ML Platform Engineer / MLOps Engineer
Technical requirement: API integration + infrastructure

Primarily work with model serving and deployment rather than training
Focus: KServe, Ray Serve, TorchServe, Triton Inference Server
Use PyTorch/TF knowledge for optimization (quantization, compilation), not training from scratch
Your fit: Strong match—your KServe contributions and Meta infrastructure experience align perfectly

2. AI Application Engineer
Technical requirement: API integration (90%)

Build products using OpenAI, Anthropic, Cohere APIs
Minimal training from scratch—mostly prompt engineering, RAG, fine-tuning via APIs
Focus: LangChain, vector databases, orchestration frameworks
Your fit: Medium—less infrastructure-focused, more product layer

3. ML Infrastructure Engineer
Technical requirement: Training from scratch + infrastructure

Build distributed training systems, GPU clusters, data pipelines
Deep PyTorch knowledge required for custom training loops, distributed strategies
Focus: PyTorch Distributed, DeepSpeed, Megatron, FSDP
Your fit: Strong match—combines your production engineering background with ML systems

4. AI Safety/Alignment Engineer
Technical requirement: Training from scratch (70%) + API integration (30%)

RLHF, constitutional AI, red-teaming require custom training loops
Some work with API-based evaluation and testing
Focus: Custom PyTorch implementations of alignment techniques
Your fit: Lower—requires more research orientation

5. Inference Optimization Engineer
Technical requirement: Neither pure training nor pure API—specialized optimization

Optimize pre-trained models for production (quantization, pruning, distillation)
Requires PyTorch understanding but not training from scratch
Focus: ONNX, TensorRT, OpenVINO, custom CUDA kernels
Your fit: Strong match—ties to your inference-perf contributions and production focus

Key Pattern for Your Background:
Roles #1, #3, and #5 are growing fastest because companies have models but need production infrastructure. Your Meta production engineering + KServe/Kubernetes work positions you perfectly for the infrastructure side of AI, not the training/research side.
Strategic insight: "Training from scratch" roles are actually growing slower than infrastructure roles—most companies fine-tune or use pre-trained models. The bottleneck is deployment, scaling, and serving.
```

In [58]:
%%html
<div class="todo-box">
<h3>TODO</h3>
</div>

## Part 2: The "Deep Search" Verification


Now that you have the platforms and titles, go find the data.

1. Go to one of the platforms your AI suggested (e.g., **Otta**, **Wellfound**, **Hiring Cafe**? There are lots!)
2. Find **two** job postings that are of interest to you.
* One for a **"Traditional Software Engineer"** or similar entry level title
* One for an **"AI Engineer"** or **"ML Engineer"** or similar title (there are a lot of titles like these)
3. **The Analysis:** List the actual title used, three specific technical skills (not soft skills) related to AI/ML required for the role that appear in the posting.Also list the years of experience required, if any.


#### ANSWER:

1. AI Engineer @ Eudia [link](https://wellfound.com/jobs/3714087-ai-engineer) - 2-5 years of experience

- Proficiency in LLMs, RAGs, evaluation techniques, and machine learning frameworks (e.g., TensorFlow, PyTorch) and programming languages such as Python.
- Familiarity with MLOps practices and tools (e.g., MLflow) and experience in model deployment, monitoring, and continuous integration/continuous deployment (CI/CD) of AI/ML solutions.
- Experience with cloud platforms (e.g., AWS, Azure, Google Cloud) and ML services is essential.


2. Software Engineer @ Agave API [link](https://wellfound.com/jobs/3272257-software-engineer) - 1-4 years of experience

- Back-end/full-stack engineer, ideally with 1-4 years experience (we're open to exceptional college grads with strong intern/co-op experience).
- Familiarity with cloud infrastructure (e.g. AWS).
- TypeScript, React, Redis, Postgres, AWS - CDK, EC2

In [59]:
%%html
<div class="todo-box">
<h3>TODO</h3>
</div>

## Part 3: Decoding the "AI Engineer" Title

**Mandatory Reading:** [The Rise of the AI Engineer (swyx.io)](https://www.google.com/search?q=https://www.swyx.io/ai-engineer) The site https://swyx.io is a good site for you to bookmark. Shawn Wang has received a lot of press in the AI world for his work in raising awareness of the opportunities with AI in the tech field. When you go to his site, look for his article that really caused him to take off in popularity: https://www.latent.space/p/ai-engineer. It's 2023, so outdated on the AI timeline, except in this case. It's still very relevant reading. 

*(Note to Students: This is the foundational text that defined the shift from ML Researchers to AI Engineers. Read it carefully.)*

**Reflection:** Based on your job search in Part B and this reading, do you see the market shifting more toward "Researchers" who build models, or "Engineers" who build *with* models? Which path interests you more and why?


#### ANSWER: 

I see the market shifting more toward "Engineers" who build with models because that is where humans are able to extract value from machine learning models. Without the engineers, ML does not provide utility to the general population. The "Engineer" path definitely interests me more as I feel like it will allow me to have a more direct impact on people's lives. I want to be closer to designing and creating products (or at least help design infrastructure to support this).




---

In [60]:
%%html
<div class="section-box">
<h1></h1>
</div>

# <div class="section-box">Society</div>

## The Global Hardware Divide

**Objective:** To understand the environmental and geopolitical "footprint" of the hardware that makes PyTorch possible.

### The Context

As a student at Bucknell, you have access to high-performance systems. Many of you have some pretty good systems back at your dorm! Some of you have high end laptops with 4000 or even 5000 NVidia RTX GPUs, or Macbook Pros with M5 processors! You have free cloud GPUs (like Google Colab T4s). The "AI Revolution" is not evenly distributed. Not even close. While we focus on the math of the neurons, a global infrastructure of mining, manufacturing, and energy consumption is struggling to keep up with AI demands.


In [61]:
%%html
<div class="todo-box">
<h3>TODO</h3>
</div>

## Reading: AI Boom and CO2 Emissions

Read the following article from **The Guardian** (published December 2025) which details the massive environmental surge caused by the AI boom.

* **URL:** [AI boom has caused same CO2 emissions in 2025 as New York City, report claims](https://www.theguardian.com/technology/2025/dec/18/2025-ai-boom-huge-co2-emissions-use-water-research-finds)

In this assignment, you are running code on a GPU. We ought not to refer to using a GPU as a free resource! It has a real physical cost in water and carbon. This article cites the latest 2025 data showing that AI-related carbon emissions now rival major global cities and that water usage for cooling data centers has exceeded global bottled-water demand.

Write a **250-word (or less) reflection** addressing the following two points. Be prepared to discuss them in class.

1. **The Extraction Cost:** The article mentions that "society is paying for these costs, not the tech companies." Do you agree? As a future developer, if you build a popular AI app that requires millions of PyTorch inferences daily, do you believe you have a "carbon debt" to the public? How would you design your code differently to minimize this?
2. **The Hardware Privilege:** Most AI progress is driven in the Global North, yet the environmental and labor "externalities" (mining, e-waste, energy strain) often fall on the Global South. Does this "Hardware Divide" mean that only wealthy nations get to decide the "rules" of AI?

These reflections only require you to think like a **responsible engineer**.


#### ANSWER:

Yes, I do agree that "society is paying for these costs, not the tech companies". However, I would like to also argue that all types of industries in the world are doing similar things to some extent. For instance, if we take a look at the pharmaceutical industry, we can also see that they still emit a lot of waste. This waste is then paid for by society as it harms the natural environment that people live in. Every industry on Earth will have some form of waste production at some scale. The only thing that differentiates the AI industry is the scale at which this waste is produced. This is why I believe that ALL industries have a "waste debt" to the public. Designing code differently to minimize this may not be impactful enough to mitigate the damage being done. I think that society should focus our efforts on exploring cleaner ways of generating electricity and manufacturing, such as nuclear fusion (though this may be being prevented by the oil magnates).

Yes, I believe that only wealthy nations will get to decide the "rules" of AI for the most part. The reason that I believe this is that AI is now viewed as a nation-level priority just as weapons and rocket science were during the Cold War. Due to this level of precedence, many nations who are already in the lead of AI development and research will not want to lose the lead, leaving many of the less-developed nations behind.




---

In [62]:
%%html
<div class="section-box">
<h1></h1>
</div>

# <div class="section-box">AI Usage Disclosure</div>

**IDE or browser-based environment used:**

VSCode? Google Colab? PyCharm? Cursor? JupyterLab? Something else?

**AI tools used:**

Provide a list of all AI tools you used while completing this work.(e.g. CoPilot? CodeGPT? PyCharm AI Assistant? Cursor’s AI? Perplexity? CodeLLM? Etc.)

**Short reflection**

How did you use AI? What value did AI contribute for you? Explain. If it was a hindrance, why? Discuss. *Be honest in this*! There is NO penalty for thoughtful AI use that is properly disclosed.

For individual problems that required heavy AI reliance for you to answer, include additional criteria at the end of your question:
**AI Usage Disclosure for Question**

-   Why was it challenging?

-   How did AI help you solve the problem?

-   What did you learn by using AI on this problem?

-   How might you approach a similar problem in the future? Do you feel you could solve it independently now?

**AI Chat Dialogs**

-   YOU MUST COPY AND PASTE ALL RELEVANT CHAT DIALOGS THAT PLAYED A CRITICAL ROLE IN HELPING YOU COMPLETE YOUR WORK. Please redact any part of the chat that is not relevant to this work.



<b>IDE</b>: VSCode with Colab

<b>AI Tools Used</b>: Claude

<b>Short Reflection</b>: I used AI to help find pytorch functions and syntax. It was not a hindrance, but it helped me save time from looking through documentation (for quick search).

<b>AI Chat Dialogues</b>:

USER: how do i calculate the fraction of tensors that have values more than 1.0

CLAUDE: Use a boolean comparison, convert to float, then take the mean.Method 1: Boolean mask → mean (cleanest)pythontensor = dist.sample((10000,))

fraction = (tensor > 1.0).float().mean()
print(f"Fraction > 1.0: {fraction:.4f}")